### PR TITLE
Add on_setattr hooks to attr.s and attr.ib

### LINKED ISSUE #645: [RFC] __setattr__ hooks
There's two things that people keep asking for:

1. validation on setting attributes
2. freezing single attributes

Those two features have something in common: they require `attrs` to write a `__setattr__` method.

I actually had 1 done when I implemented validators but I took it out again, because I didn't want to tamper with `__setattr__` too. But it totally makes sense to expect that validators run there too.

Now that argument has gone away thanks to frozen classes and `attrs` is in the `__setattr__` business. So it feels like the right thing to do, to implement it and make it default for _Operation `import attrs`_ (I hope this is legit the last part of the puzzle).

To allow for 2 too, I would suggest to add a hook called `on_setattr` (better names welcome) that takes a callable that is called with the instance, the attribute definition, and the new value.

To solve 2, the implementation would look like

```python
def frozen(_, __, ___):
    raise FrozenInstanceError
```

***

Open questions:

- what to do about `on_setattr` attributes in a frozen class (incl inheritance)
- what about converters? Maybe it should take a list/`and` like validator/converter do? They would need to work as a chain, returning values for the next one.

### PR DESCRIPTION
**Pull request by hynek**
This fixes #645 and a billion other issues that asked for freezing attributes and validation on assignment.

I'm not married to all the names and open to better suggestions.

Yes, it was more work to integrate than anticipated. 😳

---

While working on this I've also noticed that we take no precaution about `__setattr__` being inherited, which means that if you subclass a frozen class, you can't easily thaw it. This topic is somewhat related to this PR too but I didn't want to bloat it any further. It goes down to setting `__setattr__` to `object.__setattr__` if we don't write an own, but if `auto_detect` is true, only if the user didn't write one?

### CODE REVIEW

File: src/attr/setters.pyi [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+from typing import TypeVar, Any, NewType, cast
+
+_T = TypeVar("_T")
+
+def pipe(*setters: _OnSetAttrType) -> _OnSetAttrType: ...
+def frozen(instance: Any, attribute: Attribute, new_value: Any) -> Any: ...
+def validate(instance: Any, attribute: Attribute, new_value: _T) -> _T: ...
+def convert(instance: Any, attribute: Attribute, new_value: _T) -> _T: ...
+
+DisableType = NewType("DisableType", object)
```
Comment (euresti) [inline src/attr/setters.pyi:11]: You might want to warn that this isn't a runtime type.  Maybe even make it `_DisableType` to enforce that.

File: changelog.d/645.change.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+It is now possible to specify hooks that are called whenever an attribute is set **after** a class has been instantiated.
```
Comment (euresti) [inline changelog.d/645.change.rst:1]: Is this file necessary?

File: tests/test_functional.py [hunk 3]
```diff
@@ -681,3 +683,183 @@ class C(object):
...
+            y = attr.ib()
+
+        pf = PartiallyFrozen("x", "y")
+
+        pf.y = "yyy"
+
+        assert "yyy" == pf.y
+
+        with pytest.raises(FrozenAttributeError):
+            pf.x = "xxx"
```
Comment (euresti) [inline tests/test_functional.py:731]: Just FYI.  Mypy will not catch this as an error.  Not important for this diff but something to consider.

File: src/attr/__init__.pyi [hunk 4]
```diff
@@ -189,6 +199,7 @@ def attrs(
     order: Optional[bool] = ...,
     auto_detect: bool = ...,
     getstate_setstate: Optional[bool] = ...,
+    on_setattr: Optional[_OnSetAttrArgType] = ...,
```
Comment (euresti) [inline src/attr/__init__.pyi:202]: Does this actually work?
```
@attr.s(on_setattr=DISABLE)
```

File: src/attr/__init__.pyi [hunk 5]
```diff
@@ -41,6 +42,10 @@ _ConverterType = Callable[[Any], _T]
 _FilterType = Callable[[Attribute[_T], _T], bool]
 _ReprType = Callable[[Any], str]
 _ReprArgType = Union[bool, _ReprType]
+_OnSetAttrType = Callable[[Any, Any, Any], Any]
```
Comment (euresti) [inline src/attr/__init__.pyi:45]: Could this be:
```
Callable[[Any, Attribute[_T], _T], _T]
```
?
It might just be enough to do:
```
Callable[[Any, Attribute[Any], Any], Any]
```

File: src/attr/setters.pyi [hunk 6]
```diff
@@ -0,0 +1,12 @@
+from . import _OnSetAttrType, Attribute
+from typing import TypeVar, Any, NewType, cast
+
+_T = TypeVar("_T")
+
+def pipe(*setters: _OnSetAttrType) -> _OnSetAttrType: ...
+def frozen(instance: Any, attribute: Attribute, new_value: Any) -> Any: ...
+def validate(instance: Any, attribute: Attribute, new_value: _T) -> _T: ...
```
Comment (euresti) [inline src/attr/setters.pyi:8]: These could be:
```
def validate(instance: Any, attribute: Attribute[_T], new_value: _T) -> _T: ...
def convert(instance: Any, attribute: Attribute[_T], new_value: _T) -> _T: ...
```

File: src/attr/setters.pyi [hunk 7]
```diff
@@ -0,0 +1,12 @@
+from . import _OnSetAttrType, Attribute
+from typing import TypeVar, Any, NewType, cast
+
+_T = TypeVar("_T")
+
+def pipe(*setters: _OnSetAttrType) -> _OnSetAttrType: ...
+def frozen(instance: Any, attribute: Attribute, new_value: Any) -> Any: ...
```
Comment (euresti) [inline src/attr/setters.pyi:7]: I wonder if this could be:
```
from typing import NoReturn

def frozen(instance: Any, attribute: Attribute[_T], new_value: _T) -> NoReturn: ...
```

File: src/attr/__init__.pyi (hunk 4 above)
Comment (hynek) [reply src/attr/__init__.pyi:202]: It does! Mostly because I didn't want to add a sharp edge for the sake of purity:

https://github.com/python-attrs/attrs/blob/a241f0f152a99e509f572b85d4dd42f82286f1dd/src/attr/_make.py#L800

File: tests/test_functional.py (hunk 3 above)
Comment (hynek) [reply tests/test_functional.py:731]: Oh you mean mypy should be able to catch the presence of a frozen hook? That sounds tough but would be cool indeed.

File: changelog.d/645.change.rst (hunk 2 above)
Comment (hynek) [reply changelog.d/645.change.rst:1]: Unfortunately that's the only way to tell towncrier to link to another issue.

File: src/attr/__init__.pyi (hunk 5 above)
Comment (hynek) [reply src/attr/__init__.pyi:45]: What would those `_T`s buy us?

File: src/attr/setters.pyi (hunk 1 above)
Comment (hynek) [reply src/attr/setters.pyi]: Yeah, thanks!

File: src/attr/__init__.pyi (hunk 5 above)
Comment (euresti) [reply src/attr/__init__.pyi:45]: The _T means that everywhere you see it, the type must be the same.  So mypy would warn you if your method tries to change the type or if the attribute doesn't match the new value.  For example

```
def foo(instance: Any, attribute: Attribute[bool], new_value: str) -> bool:
    return new_value == 'true'

@attr.s(on_setattr=foo)
```

However looking at it, this might be something you want to do.  So I think the correct type is
```
Callable[[Any, Attribute[_T], Any], _T]
```

Basically it forces your on_setattr method to return a type that's the same as the Attribute.

Comment (hynek) [reply src/attr/__init__.pyi:45]: Is this something we want to enforce? 🤔 If you have a list, you could have intermediate results and only the last one has to return the correct one.

File: src/attr/setters.pyi (hunk 7 above)
Comment (hynek) [reply src/attr/setters.pyi:7]: NoReturn is quite fresh in typing, isn't it? I remember pulling it from mypy_extensions or something? 🤔

File: src/attr/__init__.pyi (hunk 5 above)
Comment (euresti) [reply src/attr/__init__.pyi:45]: Oh wow.  Hadn't thought of that.  Can converters be chained in the same way?

File: src/attr/setters.pyi (hunk 7 above)
Comment (euresti) [reply src/attr/setters.pyi:7]: Since this is a stub the only thing that matters is where it is in `typeshed`.

File: src/attr/__init__.pyi (hunk 5 above)
Comment (hynek) [reply src/attr/__init__.pyi:45]: They will be with the next release. At this point it's called `chain()` but I think `pipe()` is better and I will rename it before I release it.

Alternatively I could be convinced to rename pipe to chain here, but I feel like pipe is better at conveying the fact, that the data flows thru it?

File: src/attr/setters.pyi (hunk 7 above)
Comment (hynek) [reply src/attr/setters.pyi:7]: Oh!

File: src/attr/__init__.pyi (hunk 5 above)
Comment (hynek) [reply src/attr/__init__.pyi]: I went with `Attribute[Any]` for now

**Comment by hynek** [comment]
Since there was no more feedback, I'm YOLO-timeouting this. If someone wants to chime in before the release feel free.

**Comment by hynek** [comment]
P.S. @euresti I took the liberty to relax `_ConverterType` based on your comments in a74c32b…feel free to yell at me if that's terrible and should be done differently.

File: src/attr/setters.pyi [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+_T = TypeVar("_T")
+
+def frozen(
+    instance: Any, attribute: Attribute, new_value: Any
+) -> NoReturn: ...
+def pipe(*setters: _OnSetAttrType) -> _OnSetAttrType: ...
+def validate(instance: Any, attribute: Attribute[_T], new_value: _T) -> _T: ...
+
+# convert is allowed to return Any, because they cann be chained using pipe.
```
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/setters.pyi:12]: Good catch, fixed in d6ba983edb43ef04d1074109301901a1b8147a9d

**Comment by hhnguyen-axon** [comment]
Hi @hynek, I wonder if we have any release plan for this fix? We're having some use-cases which requires this and we're eagerly looking for it to be released. If anything we can help to make it release sooner, can you let us know?

**Comment by hynek** [comment]
#666 is the last remaining part of the puzzle. I hope I’ll be able to release this week.

**Comment by hhnguyen-axon** [comment]
Sounds great, thanks so much @hynek
